httpse.go

```diff
@@ -0,0 +1,157 @@
+package httpseverywhere
+
+import (
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/armon/go-radix"
+	"github.com/getlantern/golog"
+	"github.com/getlantern/mtime"
+)
+
+// Rewrite changes an HTTP URL to rewrite.
+type Rewrite func(url *url.URL) (string, bool)
+
+type httpse struct {
+	log             golog.Logger
+	initOnce        sync.Once
+	wildcardTargets atomic.Value // *radix.Tree
+	plainTargets    atomic.Value // map[string]*ruleset
+	stats           *httpseStats
+	statsCh         chan *timing
+}
+
+type httpseStats struct {
+	runs      int64
+	totalTime int64
+	max       int64
+	maxHost   string
+}
+
+type timing struct {
+	host string
+	dur  time.Duration
+}
+
+// Default returns a lazily-initialized Rewrite using the default rules
+func Default() Rewrite {
+	h := newEmpty()
+	h.initAsync()
+	return h.rewrite
+}
+
+func newEmpty() *httpse {
+	h := &httpse{
+		log:     golog.LoggerFor("httpse"),
+		stats:   &httpseStats{},
+		statsCh: make(chan *timing),
+	}
+	go h.readTimings()
+	h.wildcardTargets.Store(radix.New())
+	h.plainTargets.Store(make(map[string]*ruleset))
+	return h
+}
+
+func (h *httpse) init() {
+	d := newDeserializer()
+	plain, wildcard, err := d.newRulesets()
+	if err != nil {
+		return
+	}
+	h.plainTargets.Store(plain)
+	h.wildcardTargets.Store(wildcard)
+}
+
+func (h *httpse) initAsync() {
+	h.initOnce.Do(func() {
+		go h.init()
+	})
+}
+
+func (h *httpse) rewrite(url *url.URL) (string, bool) {
+	if url.Scheme != "http" {
+		return "", false
+	}
+
+	start := mtime.Now()
```

Comment:
You can also use `mtime.Stopwatch()` if you want, which was designed for this exact purpose.